src/Type/Php/ArrayShiftFunctionReturnTypeExtension.php

```diff
@@ -43,7 +43,8 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
 					continue;
 				}
 
-				$valueTypes[] = $constantArray->getOffsetValueType($arrayKeyTypes[0]);
+				$constantArray->removeFirst($removedValueType);
```

Comment:
wild guess: instead of doing it by-ref, I think cou could `$constantArray->count()` before and after the `removeFirst` call and determine the `removedValueType` by `TypeCombinator::union($beforeConstantArray->getValueTypes(), $afterConstantArray->getValueTypes())`

Reply:
it's a bit trickier because of optional keys, but give me a minute, I'll push a simpler version soon. Trying to reuse existing code here was a bad idea :)

Reply:
pushed, what do you think? makes more sense to me :)